File: src/cleansia_ios/CleansiaPartner/Sources/Features/Profile/SectionScaffold.swift
import CleansiaCore
import SwiftUI

struct SectionScaffold<Header: View, Form: View>: View {
    let title: String
    let isLoading: Bool
    var isError: Bool = false
    var onRetry: (() -> Void)?
    @ViewBuilder let header: () -> Header
    @ViewBuilder let form: () -> Form

    init(
        title: String,
        isLoading: Bool,
        isError: Bool = false,
        onRetry: (() -> Void)? = nil,
        @ViewBuilder header: @escaping () -> Header = { EmptyView() },
        @ViewBuilder form: @escaping () -> Form
    ) {
        self.title = title
        self.isLoading = isLoading
        self.isError = isError
        self.onRetry = onRetry
        self.header = header
        self.form = form
    }

    var body: some View {
        Group {
            if isLoading {
                VStack {
                    Spacer().frame(height: 80)
                    ProgressView()
                    Spacer()
                }
                .frame(maxWidth: .infinity)
            } else if isError, let onRetry {
                SectionErrorRetry(onRetry: onRetry)
            } else {
                ScrollView {
                    VStack(alignment: .leading, spacing: Spacing.m) {
                        header()
                        form()
                    }
                    .padding(.horizontal, Spacing.m)
                    .padding(.bottom, Spacing.l)
                }
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(CleansiaColors.background.ignoresSafeArea())
        .navigationTitle(title)
        .navigationBarTitleDisplayMode(.inline)
    }
}

private struct SectionErrorRetry: View {
    let onRetry: () -> Void

    var body: some View {
        VStack(spacing: Spacing.l) {
            Text(L10n.Profile.errorGeneric)
                .font(CleansiaTypography.bodyMedium)
                .foregroundColor(CleansiaColors.onSurfaceVariant)
                .multilineTextAlignment(.center)
            CleansiaPrimaryButton(L10n.retry, action: onRetry)
                .fixedSize()
        }
        .padding(Spacing.xl)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

File: src/cleansia_android/partner-app/src/main/java/cz/cleansia/partner/features/profile/SectionScaffold.kt
package cz.cleansia.partner.features.profile

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.ui.text.style.TextAlign
import cz.cleansia.core.ui.components.CleansiaPrimaryButton
import cz.cleansia.core.ui.theme.Spacing
import cz.cleansia.partner.R

/**
 * Shared scaffold for profile-section editors. Wraps a Material
 * `TopAppBar` (matches the customer profile's editor chrome) over a
 * scrollable form area. Errors flow through the app-wide
 * GlobalSnackbarHost — VMs push there directly so every section sees
 * the same branded toast.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SectionScaffold(
    title: String,
    isLoading: Boolean,
    onNavigateBack: () -> Unit,
    isError: Boolean = false,
    onRetry: (() -> Unit)? = null,
    headerSlot: (@Composable ColumnScope.() -> Unit)? = null,
    form: @Composable ColumnScope.() -> Unit,
) {
    Scaffold(
        topBar = {
            TopAppBar(
                title = {
                    Text(
                        text = title,
                        style = MaterialTheme.typography.titleLarge,
                    )
                },
                navigationIcon = {
                    IconButton(onClick = onNavigateBack) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Outlined.ArrowBack,
                            contentDescription = stringResource(R.string.back),
                        )
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = MaterialTheme.colorScheme.background,
                    titleContentColor = MaterialTheme.colorScheme.onBackground,
                    navigationIconContentColor = MaterialTheme.colorScheme.onBackground,
                ),
            )
        },
        containerColor = MaterialTheme.colorScheme.background,
    ) { paddingValues ->
        if (isLoading) {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(paddingValues),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Spacer(Modifier.height(80.dp))
                CircularProgressIndicator()
            }
        } else if (isError && onRetry != null) {
            // The load failed. Before, the Error state fell through to an empty editable form with no
            // way to recover; render a message + retry instead (T-0353).
            ErrorRetry(onRetry = onRetry, paddingValues = paddingValues)
        } else {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(paddingValues)
                    .background(MaterialTheme.colorScheme.background)
                    .verticalScroll(rememberScrollState())
                    .padding(horizontal = Spacing.M)
                    .padding(bottom = Spacing.L),
            ) {
                if (headerSlot != null) {
                    headerSlot()
                    Spacer(Modifier.height(Spacing.M))
                }
                form()
            }
        }
    }
}

@Composable
private fun ErrorRetry(onRetry: () -> Unit, paddingValues: PaddingValues) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(paddingValues)
            .padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Text(
            text = stringResource(R.string.error_generic),
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
        )
        Spacer(Modifier.height(Spacing.L))
        CleansiaPrimaryButton(
            text = stringResource(R.string.retry),
            onClick = onRetry,
        )
    }
}
